go/vt/sqlannotations/sqlannotations.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2015, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package sqlannotations provides functions
+// for annotating DML statements with keyspace-id
+// comments and parsing them. These annotations
+// are used during filtered-replication to route
+// the DML statement to the correct shard.
+package sqlannotations
+
+import (
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// AnnotateKeyspaceIdIfDML annotates 'sql' with the given keyspace id
+// if 'sql' is a DML statement; otherwise, it does nothing.
+func AnnotateKeyspaceIdIfDML(keyspaceId []byte, sql *string) {
+	if isDml(*sql) {
+		*sql = fmt.Sprintf("%s /* vtgate:: keyspace_id:%s */",
+			*sql, hex.EncodeToString(keyspaceId))
+	}
+}
+
+// AnnotateAsFilteredReplicationUnfriendlyIfDML annotates the given 'sql'
+// query as filtered-replication-unfriendly if its a DML statement
+// (filtered-replication will log an error if it encounters such a annotation).
+// Does nothing if the statement is not a DML statement.
+func AnnotateAsFilteredReplicationUnfriendlyIfDML(sql *string) {
+	if isDml(*sql) {
+		*sql = *sql + " /* vtgate:: filtered_replication_unfriendly */"
+	}
+}
+
+// Copied from go/vt/vtgate/resolver.go
+// TODO(erez): Refactor this.
+var (
+	insertDML = "insert"
+	updateDML = "update"
+	deleteDML = "delete"
+)
+
+// Returns true if 'querySQL' is an INSERT, UPDATE or DELETE
+// statement.
+func isDml(querySQL string) bool {
+	var sqlKW string
+	if i := strings.Index(querySQL, " "); i >= 0 {
+		sqlKW = querySQL[:i]
+	}
+	sqlKW = strings.ToLower(sqlKW)
+	if sqlKW == insertDML || sqlKW == updateDML || sqlKW == deleteDML {
+		return true
+	}
+	return false
+}
+
+var (
+	keyspaceIdReplicationRegexp         = regexp.MustCompile("/\\* vtgate:: keyspace_id:([^ ]*) \\*/")
```

Comment:
The regexp here seems a bit overkill, for just a couple string indexes... We commit 5k QPS+, so performance matters.

Would you mind running a small performance experiment to see the CPU usage of a simple string search? Can be in another change...